torch/csrc/jit/passes/python_print.cpp

```diff
@@ -167,7 +167,74 @@ const static std::unordered_set<std::string> reserved_names = {
 };
 
 struct PythonPrintPass {
-  std::ostringstream body_;
+  using SourceRangeStack = std::vector<SourceRange>;
+  SourceRangeStack source_range_stack_ = {SourceRange("")};
+
+  struct WithSourceRange {
+    explicit WithSourceRange(SourceRangeStack* stack, SourceRange sr)
+        : stack(stack) {
+      TORCH_INTERNAL_ASSERT(stack);
+      stack->push_back(std::move(sr));
+    }
+
+    ~WithSourceRange() {
+      stack->pop_back();
+    }
+
+    SourceRangeStack* stack;
+  };
+
+  class TaggedStringStream {
+   public:
+    TaggedStringStream(const SourceRangeStack* srs) : srs(srs) {}
+    TaggedStringStream(TaggedStringStream&& rhs) = default;
+
+    TaggedStringStream& operator<<(const std::string& s) {
+      ranges_.emplace_back((size_t)oss.tellp(), srs->back());
```

Comment:
This is hugely inefficient! Most of the time the source range is exactly the same as the previous one, but this takes each individual call to << and makes a new 24 byte object to record its source (before serialization, the tuples are larger). This should check that the ranges_.back() has the same source and if so just no-op.